File: Programmers/javascript/기능개발.js
/*
프로그래머스 기능개발
progresses배열의 값을 완료까지 남은 일수로 바꾸어 작업
현재 progresses값이 이전 progresses값보다 작다면 작업일수+1
작지않다면 date값 push 후 date값 초기화
*/
function solution(progresses, speeds) {
    var answer = [];
    for(let i=0; i<progresses.length; i++){
        progresses[i]=100-progresses[i];
        if(progresses[i]%speeds[i]==0){
            progresses[i]/=speeds[i];
        }else{
            progresses[i]=parseInt(progresses[i]/speeds[i])+1;
        }
    }
    //console.log(progresses);
    let start=progresses[0];
    let date=1;
    for(let i=1; i<progresses.length; i++){
        if(start>=progresses[i]){
            date++;
        }else{
            answer.push(date);
            start=progresses[i];
            date=1;
        }
    }
    answer.push(date);
    return answer;
}

File: Programmers/java/기능개발.java
/*
프로그래머스 기능개발
queue를 사용하여 해결.
task=process가 끝나는데 걸리는 일 수
현재 task가 다음 프로세스의 task보다 작다면 배포
*/
import java.util.*;
class Solution {
    public ArrayList<Integer> solution(int[] progresses, int[] speeds) {
        ArrayList<Integer> answer = new ArrayList<Integer>();
        Queue<Integer> q = new LinkedList<>();
        for(int i=0; i<progresses.length; i++){
            int task=100-progresses[i];
            if(task%speeds[i]==0) q.offer(task/speeds[i]);
            else q.offer(task/speeds[i] + 1);
        }
        int process=1;
        int task=q.poll();
        while(!q.isEmpty()){
            int fr=q.poll();
            if(task < fr){
                task=fr;
                answer.add(process);
                process=1;
            }else{
                process++;
            }
        }
        answer.add(process);
        return answer;
    }
}
